File: src/javascript/algorithms/math/sortIntegersByPowerValue/sortIntegersByPowerValue.js
function getKth(low, high, k) {
    const memo = new Map([[1, 0]]);

    function getPower(x) {
        if (memo.has(x)) return memo.get(x);
        const steps = x % 2 === 0 ? 1 + getPower(x / 2) : 1 + getPower(3 * x + 1);
        memo.set(x, steps);
        return steps;
    }

    const pairs = [];
    for (let i = low; i <= high; i++) {
        pairs.push([i, getPower(i)]);
    }
    pairs.sort((a, b) => a[1] !== b[1] ? a[1] - b[1] : a[0] - b[0]);
    return pairs[k - 1][0];
}

const r1 = getKth(12, 15, 2);
console.log(`Test 1 - lo=12 hi=15 k=2: ${r1 === 13 ? '[PASS]' : '[FAIL]'} result=${r1} expected=13`);

const r2 = getKth(7, 11, 4);
console.log(`Test 2 - lo=7 hi=11 k=4: ${r2 === 7 ? '[PASS]' : '[FAIL]'} result=${r2} expected=7`);

const r3 = getKth(1, 5, 3);
console.log(`Test 3 - lo=1 hi=5 k=3: ${r3 === 4 ? '[PASS]' : '[FAIL]'} result=${r3} expected=4`);

const r4 = getKth(3, 5, 2);
console.log(`Test 4 - lo=3 hi=5 k=2: ${r4 === 5 ? '[PASS]' : '[FAIL]'} result=${r4} expected=5`);

const r5 = getKth(1, 3, 1);
console.log(`Test 5 - lo=1 hi=3 k=1: ${r5 === 1 ? '[PASS]' : '[FAIL]'} result=${r5} expected=1`);

File: src/python/algorithms/math/sortIntegersByPowerValue/sort_integers_by_power_value.py
import sys
sys.setrecursionlimit(10000)


def power(x, memo):
    if x in memo:
        return memo[x]
    if x % 2 == 0:
        steps = 1 + power(x // 2, memo)
    else:
        steps = 1 + power(3 * x + 1, memo)
    memo[x] = steps
    return steps


def get_kth(low, high, k):
    memo = {1: 0}
    pairs = []
    for i in range(low, high + 1):
        pairs.append((i, power(i, memo)))
    pairs.sort(key=lambda x: (x[1], x[0]))
    return pairs[k - 1][0]


if __name__ == "__main__":
    r1 = get_kth(12, 15, 2)
    print(f"Test 1 - lo=12 hi=15 k=2: {'[PASS]' if r1 == 13 else '[FAIL]'} result={r1} expected=13")

    r2 = get_kth(7, 11, 4)
    print(f"Test 2 - lo=7 hi=11 k=4: {'[PASS]' if r2 == 7 else '[FAIL]'} result={r2} expected=7")

    r3 = get_kth(1, 5, 3)
    print(f"Test 3 - lo=1 hi=5 k=3: {'[PASS]' if r3 == 4 else '[FAIL]'} result={r3} expected=4")

    r4 = get_kth(3, 5, 2)
    print(f"Test 4 - lo=3 hi=5 k=2: {'[PASS]' if r4 == 5 else '[FAIL]'} result={r4} expected=5")

    r5 = get_kth(1, 3, 1)
    print(f"Test 5 - lo=1 hi=3 k=1: {'[PASS]' if r5 == 1 else '[FAIL]'} result={r5} expected=1")
